google-ads-examples/src/main/java/com/google/ads/googleads/examples/reporting/ParallelReportDownload.java

```diff
@@ -0,0 +1,253 @@
+      // library will automatically retry with exponential back-off before failing the request.
+      for (String gaqlQuery : GAQL_QUERY_STRINGS) {
+        // Uses a list of futures to make sure that we wait for this report to complete on all
+        // customer IDs before proceeding. The Future data type here is just for demonstration.
+        List<ListenableFuture<ReportSummary>> futures = new ArrayList<>();
+
+        // Calls the API for each customer ID on the current report.
+        for (Long customerId : customerIds) {
+          // Uses the gRPC asynchronous API to download the reports in parallel. This saves having
+          // to create/manage our own thread pool.
+          ResponseCountingObserver responseObserver = new ResponseCountingObserver(customerId);
+
+          // Starts the report download in a background thread.
+          serviceClient
+              .searchStreamCallable()
+              .call(
+                  SearchGoogleAdsStreamRequest.newBuilder()
+                      .setCustomerId(customerId.toString())
+                      .setQuery(gaqlQuery)
+                      .build(),
+                  responseObserver);
+
+          // Stores a future to retrieve the results.
+          futures.add(responseObserver.asFuture());
+        }
+
+        // Waits for all pending requests to the current set of customer IDs to complete.
+        //
+        // This is a naive implementation for illustrative purposes. It is possible to optimize the
+        // utilization of each customer ID by providing a queue of work (or similar). However, this
+        // would complicate the example code and so is omitted here.
+        List<ReportSummary> results =
+            Futures.allAsList(futures)
+                .get(DEADLINE_MILLIS - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+        System.out.println("Report results for query: " + gaqlQuery);
+        results.forEach(System.out::println);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException | TimeoutException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** An observer which records a simple count of the result rows received. */
+  private static class ResponseCountingObserver
+      implements ResponseObserver<SearchGoogleAdsStreamResponse> {
+
+    private final long customerId;
+    private final SettableFuture<ReportSummary> future = SettableFuture.create();
+    private final AtomicLong numResponses = new AtomicLong(0);
+
+    ResponseCountingObserver(long customerId) {
+      this.customerId = customerId;
+    }
+
+    @Override
+    public void onStart(StreamController controller) {
+      // Nothing to do here.
+    }
+
+    @Override
+    public void onResponse(SearchGoogleAdsStreamResponse response) {
+      // Does something useful with the response. In this case we just count the responses.
```

Comment:
Worth expanding a bit to something like "This implementation just counts the responses, but could also write the response to a file, pass the response on to another method for further processing, etc.."
Just want to make it clear to users that this is an intentionally simple example for readability, but they can get creative and do whatever they'd like with the results.

Reply:
sg - done